### ISSUE TITLE
YOLO11's maximum detecting categories

### ISSUE THREAD

**Issue by liuyanliang666** [ultralytics/ultralytics#21770]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I want to train yolo to detect more than 1000 types of objects. Is this possible? Is so, what can i do to improve accuracy.Now i have trained yolo to detect 36 classes, but the accuracy is relatively low.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @liuyanliang666, thank you for your interest in Ultralytics 🚀! This is an automated response to help get things started. An Ultralytics engineer will follow up with you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including:
-Your exact training/validation commands and ultralytics package version
-The YOLO11 model variant used and input image size
-Dataset details: number of classes, images per class, example images/labels, and any class imbalance
-Training logs (metrics per epoch), confusion matrix, and validation results
-Any custom augmentations or hyperparameters used
-Environment details: OS, Python, PyTorch, CUDA versions, and device
Also verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a **Python>=3.8** environment with **PyTorch>=1.8** to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Yes—YOLO can train with thousands of classes; just list all class names in your dataset YAML `names:` (there’s no hard limit beyond compute/memory). For better accuracy, try a larger model (`yolo11x.pt`), train longer and/or at higher resolution, enable multi-scale, ensure enough clean and balanced data per class (after fixing labels, delete `labels/*.cache`). Example:

```bash
yolo detect train data=your_data.yaml model=yolo11x.pt imgsz=640 epochs=200 multi_scale=True
```

For large-vocabulary baselines see the [LVIS dataset guide](https://docs.ultralytics.com/datasets/detect/lvis/); if your taxonomy is very large or evolving, consider open-vocabulary options like [YOLO-World](https://docs.ultralytics.com/models/yolo-world/) or [YOLOE](https://docs.ultralytics.com/models/yoloe/). General training tips are in the [Object Detection docs](https://docs.ultralytics.com/tasks/detect/) and the [Troubleshooting guide](https://docs.ultralytics.com/guides/yolo-common-issues/).

**Comment by liuyanliang666** [comment]
> Yes—YOLO can train with thousands of classes; just list all class names in your dataset YAML `names:` (there’s no hard limit beyond compute/memory). For better accuracy, try a larger model (`yolo11x.pt`), train longer and/or at higher resolution, enable multi-scale, ensure enough clean and balanced data per class (after fixing labels, delete `labels/*.cache`). Example:
> 
> yolo detect train data=your_data.yaml model=yolo11x.pt imgsz=640 epochs=200 multi_scale=True
> For large-vocabulary baselines see the [LVIS dataset guide](https://docs.ultralytics.com/datasets/detect/lvis/); if your taxonomy is very large or evolving, consider open-vocabulary options like [YOLO-World](https://docs.ultralytics.com/models/yolo-world/) or [YOLOE](https://docs.ultralytics.com/models/yoloe/). General training tips are in the [Object Detection docs](https://docs.ultralytics.com/tasks/detect/) and the [Troubleshooting guide](https://docs.ultralytics.com/guides/yolo-common-issues/).

Thank you for your recommendations ! But I got another question : should I train 1000 classes at once or use transfer learning by freezing weights of the front layers?

**Comment by glenn-jocher** [comment]
Train all 1000 classes jointly from pretrained YOLO11; don’t freeze the backbone unless your dataset is very small—freezing usually hurts adaptation on large, diverse taxonomies. If you’re expanding from your 36-class model, you can initialize from it; the detection head will be reinitialized to the new `nc` automatically:
```bash
yolo detect train data=your_1k.yaml model=your_36cls.pt imgsz=640 epochs=300
```
If you still want a cautious start, you can freeze a few early layers for a short warmup using the `freeze` argument, then unfreeze and continue training. See the [Train docs](https://docs.ultralytics.com/modes/train/) for `freeze` and transfer learning details, and the [training tips](https://docs.ultralytics.com/guides/model-training-tips/) for practical advice like `batch=-1` (auto batch) and AMP.

**Comment by liuyanliang666** [comment]
> Train all 1000 classes jointly from pretrained YOLO11; don’t freeze the backbone unless your dataset is very small—freezing usually hurts adaptation on large, diverse taxonomies. If you’re expanding from your 36-class model, you can initialize from it; the detection head will be reinitialized to the new `nc` automatically:
> 
> yolo detect train data=your_1k.yaml model=your_36cls.pt imgsz=640 epochs=300
> If you still want a cautious start, you can freeze a few early layers for a short warmup using the `freeze` argument, then unfreeze and continue training. See the [Train docs](https://docs.ultralytics.com/modes/train/) for `freeze` and transfer learning details, and the [training tips](https://docs.ultralytics.com/guides/model-training-tips/) for practical advice like `batch=-1` (auto batch) and AMP.

Ok , but I have a problem right now . I have 2 very similar categories , which often confuse the model, what should I do to finetune the model , increase the dataset of these 2 categories or something else?

**Comment by glenn-jocher** [comment]
Common with fine-grained pairs—first verify label consistency and that the two classes are visually separable (merge if they aren’t); otherwise add more disambiguating samples for those classes, train at higher resolution, and slightly emphasize classification while turning off heavy mosaic near the end:
```bash
yolo detect train data=your.yaml model=yolo11x.pt imgsz=960 epochs=300 cls=1.0 close_mosaic=10
```
For stubborn cases, use a detector+classifier cascade by training a small 2-class image classifier on detection crops; see the [Classification task docs](https://docs.ultralytics.com/tasks/classify/), and track improvements with the confusion matrix in the [Val docs](https://docs.ultralytics.com/modes/val/).

**Comment by liuyanliang666** [comment]
> Common with fine-grained pairs—first verify label consistency and that the two classes are visually separable (merge if they aren’t); otherwise add more disambiguating samples for those classes, train at higher resolution, and slightly emphasize classification while turning off heavy mosaic near the end:
> 
> yolo detect train data=your.yaml model=yolo11x.pt imgsz=960 epochs=300 cls=1.0 close_mosaic=10
> For stubborn cases, use a detector+classifier cascade by training a small 2-class image classifier on detection crops; see the [Classification task docs](https://docs.ultralytics.com/tasks/classify/), and track improvements with the confusion matrix in the [Val docs](https://docs.ultralytics.com/modes/val/).

So thankful for your reply ! I will give it a try.

**Comment by glenn-jocher** [comment]
Sounds good—when you evaluate, run the validation with plots to inspect per-class confusion and PR curves:
```bash
yolo detect val model=path/to/best.pt data=your.yaml imgsz=960 plots=True
```
For label consistency tips that help with fine-grained classes, see the [Data Collection and Annotation guide](https://docs.ultralytics.com/guides/data-collection-and-annotation/). Please share results here if you need more help, otherwise feel free to close the issue.

**Comment by liuyanliang666** [comment]
> Sounds good—when you evaluate, run the validation with plots to inspect per-class confusion and PR curves:
> 
> yolo detect val model=path/to/best.pt data=your.yaml imgsz=960 plots=True
> For label consistency tips that help with fine-grained classes, see the [Data Collection and Annotation guide](https://docs.ultralytics.com/guides/data-collection-and-annotation/). Please share results here if you need more help, otherwise feel free to close the issue.

Sorry to disturb, when trying to add pictures of  objects that have rotated 180 degrees . Should I directly take the pictures of rotated objects, or use "degree" to augment when training, which is better?

**Comment by glenn-jocher** [comment]
Real images are best if you expect upside‑down objects in production; use augmentation as a complement—prefer `flipud`/`fliplr` for 180° (less distortion) and keep `degrees` small for natural jitter. Example:
```bash
yolo detect train data=your.yaml model=yolo11x.pt imgsz=960 epochs=300 fliplr=0.5 flipud=0.5 degrees=10
```
See augmentation options in the [Train docs](https://docs.ultralytics.com/modes/train/#augmentation-settings-and-hyperparameters). If orientation itself matters, consider the [OBB task](https://docs.ultralytics.com/tasks/obb/).

**Comment by liuyanliang666** [comment]
Thank you for your suggestions! Now our goal of the task has changed, instead of detecting over 1000 categories of objects ,  i choose to combine some similar objects into one category , for example , the red flowers and yellow flowers belong to the same categories . The objects from the same categories are a bit different , but they are generally the same . For this task (which has approximately 120 categories), can I use yolo11s.pt or yolo11m.pt and train jointly instead of using transfer learning? And can yolo be able to categorize some different objects into one category ? Looking for your reply.

> Real images are best if you expect upside‑down objects in production; use augmentation as a complement—prefer `flipud`/`fliplr` for 180° (less distortion) and keep `degrees` small for natural jitter. Example:
> 
> yolo detect train data=your.yaml model=yolo11x.pt imgsz=960 epochs=300 fliplr=0.5 flipud=0.5 degrees=10
> See augmentation options in the [Train docs](https://docs.ultralytics.com/modes/train/#augmentation-settings-and-hyperparameters). If orientation itself matters, consider the [OBB task](https://docs.ultralytics.com/tasks/obb/).

**Comment by glenn-jocher** [comment]
Yes—merge the similar objects into a single class (update your YAML `names:` and relabel), and YOLO will learn the intra-class variation; for ~120 classes prefer starting from COCO‑pretrained `yolo11m.pt` (use `yolo11s.pt` only if latency is critical), then after relabeling delete `labels/*.cache` and retrain:
```bash
yolo detect train data=your_120.yaml model=yolo11m.pt imgsz=640 epochs=200 batch=-1
```
For YAML/class setup see the [Datasets guide](https://docs.ultralytics.com/datasets/), and for model size trade‑offs see the [YOLO11 docs](https://docs.ultralytics.com/models/yolo11/).

**Comment by liuyanliang666** [comment]
> Yes—merge the similar objects into a single class (update your YAML `names:` and relabel), and YOLO will learn the intra-class variation; for ~120 classes prefer starting from COCO‑pretrained `yolo11m.pt` (use `yolo11s.pt` only if latency is critical), then after relabeling delete `labels/*.cache` and retrain:
> 
> yolo detect train data=your_120.yaml model=yolo11m.pt imgsz=640 epochs=200 batch=-1
> For YAML/class setup see the [Datasets guide](https://docs.ultralytics.com/datasets/), and for model size trade‑offs see the [YOLO11 docs](https://docs.ultralytics.com/models/yolo11/).

Is imgsz=960 better ?

**Comment by glenn-jocher** [comment]
It depends—`imgsz=960` can boost small‑object recall and fine‑grained classes but costs ~2.25× more compute than 640; pick the largest size that fits your memory/latency budget and compare directly:
```bash
yolo detect val model=path/to/best.pt data=your_120.yaml imgsz=640
yolo detect val model=path/to/best.pt data=your_120.yaml imgsz=960
```
For a quick speed/accuracy profile at each size, see the CLI in the [Benchmark mode docs](https://docs.ultralytics.com/modes/benchmark/).

**Comment by liuyanliang666** [comment]
Another question. In some images , some objects aren't labeled , but they should be labeled , does it have any effect on training yolo ?

**Comment by glenn-jocher** [comment]
Yes—unlabeled in‑scope objects are learned as background, which suppresses detections and lowers recall; fully annotate every instance (or remove those images), then delete `labels/*.cache` and retrain—see “Label consistency” in the [Best YOLO Training Tips](https://docs.ultralytics.com/yolov5/tutorials/tips_for_best_training_results/) and cache behavior in the [dataset label loader reference](https://docs.ultralytics.com/reference/data/dataset/).

**Comment by liuyanliang666** [comment]
> Yes—unlabeled in‑scope objects are learned as background, which suppresses detections and lowers recall; fully annotate every instance (or remove those images), then delete `labels/*.cache` and retrain—see “Label consistency” in the [Best YOLO Training Tips](https://docs.ultralytics.com/yolov5/tutorials/tips_for_best_training_results/) and cache behavior in the [dataset label loader reference](https://docs.ultralytics.com/reference/data/dataset/).

Is there any way that can handle the problem of imbalanced data per class without manually adding the pictures ,can data augmentation help?

**Comment by glenn-jocher** [comment]
Yes—augmentation helps, but the biggest gain is rebalancing sampling; enable focal loss and stronger blends, and oversample minority images without adding new data:
```bash
yolo detect train data=your.yaml model=yolo11m.pt epochs=200 imgsz=640 mosaic=1.0 mixup=0.2 cutmix=0.1 fl_gamma=2.0 cls=1.0
```
If you can, pretrain or mix in a more diverse source to reduce long‑tail bias (see the [Objects365 dataset guide](https://docs.ultralytics.com/datasets/detect/objects365/)).

**Comment by liuyanliang666** [comment]
> Yes—augmentation helps, but the biggest gain is rebalancing sampling; enable focal loss and stronger blends, and oversample minority images without adding new data:
> 
> yolo detect train data=your.yaml model=yolo11m.pt epochs=200 imgsz=640 mosaic=1.0 mixup=0.2 cutmix=0.1 fl_gamma=2.0 cls=1.0
> If you can, pretrain or mix in a more diverse source to reduce long‑tail bias (see the [Objects365 dataset guide](https://docs.ultralytics.com/datasets/detect/objects365/)).

Can I use image_weights=True to achieve the effect of weighted sampling?

**Comment by glenn-jocher** [comment]
Not directly—YOLO11 doesn’t support `image_weights`; that flag was for older YOLO versions—see the supported training args in the Configuration docs. [Configuration (training args)](https://docs.ultralytics.com/usage/cfg/). ([docs.ultralytics.com](https://docs.ultralytics.com/usage/cfg/))

To get weighted sampling without adding new images, point `train:` to a `.txt` list and duplicate minority‑class image paths (or implement a custom sampler); the dataset loader accepts directory paths or file lists. [Object Detection Datasets Overview](https://docs.ultralytics.com/datasets/detect/). ([docs.ultralytics.com](https://docs.ultralytics.com/datasets/detect/))

```yaml
# your_data.yaml
path: /data/your_data
train: train.txt
val: val.txt
names:
  0: classA
  1: classB
```

```text
# train.txt (repeat minority-class images to oversample)
 /data/your_data/images/img001.jpg
 /data/your_data/images/minority_01.jpg
 /data/your_data/images/minority_01.jpg
 /data/your_data/images/minority_02.jpg
 /data/your_data/images/minority_02.jpg
```
